README.md
## Description

SmoothUV2 is a spatial derainbow filter. The luma is returned unchanged. This is [a port of the VapourSynth plugin SmoothUV](https://github.com/dubhater/vapoursynth-smoothuv).

SSHiQ2 is general purpose spatial denoiser. This an updated version of SSHiQ from SmoothUV v1.4.0.

### Requirements:

- AviSynth+ 3.4 or later

- Microsoft VisualC++ Redistributable Package 2022 (can be downloaded from [here](https://github.com/abbodi1406/vcredist/releases)) (Windows only)

### Usage:

```
SmoothUV2 (clip, int "radius", int "threshold", int "interlaced", int "opt", int "dither")
```

### Parameters:

- clip\
    A clip to process. It must be in YUV 8..16-bit planar format and must have at least three planes.
    
- radius\
    Must be between 1 and 7 for 4:2:0 subsampling and between 1 and 3 for 4:2:2/4:4:4 subsampling.\
    Larger values smooth more.\
    Default: 3.
    
- threshold\
    Must be between 0 and 450.\
    Larger values smooth more.\
    Default: 270.
    
- interlaced\
    Whether the frame is interlaced.\
    -1: If frame properties are supported and frame property "_FieldBased" exists - "_FieldBased" value is used.\
    If frame properties aren't supported or there is no property "_FieldBased" - 0.\
    0: Progressive frame.\
    1: Interlaced frame.\
    Default: -1.
    
- opt\
    Sets which cpu optimizations to use.\
    -1: Auto-detect.\
    0: Use C++ code.\
    1: Use SSE2 code.\
    2: Use SSSE3 code.\
    3: Use SSE4.1 code.\
    4: Use AVX2 code.\
    5: Use AVX-512 code.\
    Default: -1.
    
- dither\
    How to convert the internal 16-bit calculation to the clip depth (when clip depth < 16-bit).\
    -1: Rounding.\
    0: Ordered dither.\
    1: Error diffusion (Floyd-Steinberg).\
    Default: -1.
    
---

```
SSHiQ2 (clip, int "rY", int "rC", int "tY", int "tC", int "strY", int "strC", bool "HQY", bool "HQC", int "interlaced", int "opt", int "dither")
```

### Parameters:

- clip\
    A clip to process. It must be in YUV 8..16-bit planar format and must have at least three planes.
    
- rY, rC\
    Radius respectively for luma and chroma planes.\
    Controls how far from the current pixel neighbors are used.\
    rY Must be between 1 and 7.\
    rC Must be between 1 and 7 for 4:2:0 subsampling and between 1 and 3 for 4:2:2/4:4:4 subsampling.\
    Larger values smooth more.\
    Default: rY = 5; rC = 3.
    
- tY, tC\
    Threshold respectively for luma and chroma planes.\
    Must be between 0 and 450.\
    Set to 0 for copying the plane.\
    Larger values smooth more.\
    Default: tY = 20; tC =30.
    
- strY, strC\
    Strength respectively for luma and chroma planes.\
    It is only used for HiQ mode and it controls how much of the smoothed version should be used at best for calculating the new pixel.\
    Must be between 0 and 255.\
    Default: strY = 240; strC = strY.
    
- HQY, HQC\
    Enables HiQ mode.\
    Edge processing strength gets automatically reduced.\
    Default: HQY = HQC = true.
    
- interlaced\
    Whether the frame is interlaced.\
    -1: If frame properties are supported and frame property "_FieldBased" exists - "_FieldBased" value is used.\
    If frame properties aren't supported or there is no property "_FieldBased" - 0.\
    0: Progressive frame.\
    1: Interlaced frame.\
    Default: -1.
    
- opt\
    Sets which cpu optimizations to use.\
    -1: Auto-detect.\
    0: Use C++ code.\
    1: Use SSE2 code.\
    2: Use SSSE3 code.\
    3: Use SSE4.1 code.\
    4: Use AVX2 code.\
    5: Use AVX-512 code.\
    Default: -1.
    
- dither\
    How to convert the internal 16-bit calculation to the clip depth (when clip depth < 16-bit).\
    -1: Rounding.\
    0: Ordered dither.\
    1: Error diffusion (Floyd-Steinberg).\
    Default: -1.

### Building:

- Windows\
    Use solution files.

- Linux
    ```
    Requirements:
        - Git
        - C++17 compiler
        - CMake >= 3.16
    ```
    ```
    git clone https://github.com/Asd-g/AviSynth-SmoothUV2 && \
    cd AviSynth-SmoothUV2 && \
    mkdir build && \
    cd build && \
    cmake .. && \
    make -j$(nproc) && \
    sudo make install
    ```


CHANGELOG.md
##### 4.4.1:
    Fixed a left shift of pixel count. (dubhater)

##### 4.4.0:
    Added AVX2 and AVX512 code.

##### 4.3.1:
    Added SSSE3 code.

##### 4.3.0:
    Added SSE2 code.
    Parameter str replaced by strY and strC.

##### 4.2.1:
    Improved speed for opt=0.

##### 4.2.0:
    Added C++ code.
    Added paramters opt and dither.
    Changed internal calculation to 16-bit.
    Dropped support for AviSynth 2.x.

##### 4.1.0:
    SmoothUV2 threshold changed back to 0..450 range.
    A bit improved calculation for 8-bit.
    Added SSHiQ2 function: updated version of SSHiQ from SmoothUV v1.4.0:
    - replaced MMX asm code with SSE4.1 intrinsics;
    - fixed planes shift;
    - added support for 10..16-bit;
    - added support for 422/444 chroma subsampling;
    - more precise 8-bit calculation.

##### 4.0.0:
    Function name changed to SmoothUV2.
    Fixed horizontal planes shift.
    Fixed vertical planes shift for interlaced frames.
    Fixed output for 422/444 subsampling.
    Additional limit radius for 422/444.

##### 3.0.1:
    Another attemp to fix last columns processing.
    Removed redunant interlaced value 2.

##### 3.0.0:
    Fixed not processed edges.
    Added support for 10..16-bit.
    Set MT mode: MT_NICE_FILTER.
    Changed the range of parameter threshold from 0..450 to 0..255.
    Changed the type of parameter interlaced from bool to int.
    Changed the minimum CPU instructions support from SSE2 to SSE4.1.
    Added Linux building option.

##### 2.1.2:
    Throw error for non-planar formats.

##### 2.1.1:
    Fixed memory misalignment for AviSynth 2.6.
    Fixed processing when interlaced and h % 2.

##### 2.1.0:
    Added support for v8 interface.

##### 2.0.0:
    Port of the VapourSynth plugin SmoothUV v2.


smoothuv_5F25_dll_20030902/SmoothIQ.cpp
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <math.h>

#include "avisynth.h"
#include "SmoothFunc.h"

void (*smoothY) (int N, const unsigned char *src, unsigned char *dst, const int src_pitch,
				const int dst_pitch, int width, int height, const int t,
				unsigned short *div, unsigned char str);
void (*smoothC) (int N, const unsigned char *src, unsigned char *dst, const int src_pitch,
				const int dst_pitch, int width, int height, const int t,
				unsigned short *div, unsigned char str);

class SSmooth : public GenericVideoFilter {
private:
	bool			interlaced;
	int				DY, DC;
	int				Yth, Cth;
	unsigned short	divin[256];
	unsigned char	strength;
public:
	
	SSmooth(PClip _child, int radY, int radC, int thY, int thC, int str, bool hqy,
			bool hqc, bool field, IScriptEnvironment* env) : GenericVideoFilter(_child),
			Yth(thY), Cth(thC), interlaced(field)
	{
		if (!(env->GetCPUFlags() & CPUF_MMX))
		{
			if (thY == -1337) env->ThrowError("SmoothUV: requires a MMX-capable CPU");
			else env->ThrowError("SSHiQ: requires a MMX-capable CPU");
		}
		if (str<0 || str>255)
			env->ThrowError("SSHiQ: invalid strength");
		strength = str;

		DY = (radY<<1)+1;
		DC = (radC<<1)+1;
		if (DY<3 || DY>11)
			env->ThrowError("SSHiQ: invalid luma radius %",radY);
		if (DC<3 || DC>11)
		{
			if (thY != -1337) env->ThrowError("SSHiQ: invalid chroma radius %",radC);
			else env->ThrowError("SmoothUV: invalid chroma radius %",radC);
		}

		if (hqy) smoothY = smoothHQN_MMX;
		else smoothY = smoothN_MMX;


		if (hqc) smoothC = smoothHQN_MMX;
		else smoothC = smoothN_MMX;

		if (field)
		{
			if (thY > 0)
			{
				if (hqy) smoothY = smoothHQN_field_MMX;
				else smoothY = smoothN_field_MMX;
			}
			if (thC > 0)
			{
				if (hqc) smoothC = smoothHQN_field_MMX;
				else smoothC = smoothN_field_MMX;
			}
		}

		if (thY>450) env->ThrowError("SSHiQ: invalid Y threshold");
		if (thY == -1337)
		{
			if (thC<0)
				env->ThrowError("SmoothUV: bad thresholds, no-op filter");
			if (thC>450)
				env->ThrowError("SmoothUV: invalid chroma threshold");
		}
		/* We'll suppose the CPU hasn't changed since the start of the function */

		if (thC<=0)
		{
			if (env->GetCPUFlags() & CPUF_INTEGER_SSE) smoothC = asm_BitBlt_ISSE;
			else smoothC = asm_BitBlt_MMX;
		}
		if (thY<=0)
		{
			if (env->GetCPUFlags() & CPUF_INTEGER_SSE) smoothY = asm_BitBlt_ISSE;
			else smoothY = asm_BitBlt_MMX;
		}

		if (thC>450)
		{
			if (thY=-1337) env->ThrowError("SmoothUV: invalid chroma threshold");
			else env->ThrowError("SSHiQ: invalid U threshold");
		}
		
		/* 0 should be impossible to reach... though ...
		   Radii >= 5 are busted for now */
		divin[0] = 32767;
		for (unsigned short i=1; i<15*15; i++)
			divin[i] = (unsigned short)floor( 32767.0/i + 0.5 );

	}
	PVideoFrame __stdcall GetFrame(int n, IScriptEnvironment* env);
};

PVideoFrame __stdcall SSmooth::GetFrame(int n, IScriptEnvironment* env)
{
	PVideoFrame 		src = child->GetFrame(n, env);
	PVideoFrame 		dst = env->NewVideoFrame(vi);
	
	const int			row_size = src->GetRowSize(PLANAR_Y);
	const int			height = src->GetHeight(PLANAR_Y);
	
	smoothY(DY, src->GetReadPtr(PLANAR_Y), dst->GetWritePtr(PLANAR_Y),
			src->GetPitch(PLANAR_Y), dst->GetPitch(PLANAR_Y),
			row_size, height, Yth, divin, strength);
	smoothC(DC, src->GetReadPtr(PLANAR_U), dst->GetWritePtr(PLANAR_U),
			src->GetPitch(PLANAR_U), dst->GetPitch(PLANAR_U),
			row_size>>1, height>>1, Cth, divin, strength);
	smoothC(DC, src->GetReadPtr(PLANAR_V), dst->GetWritePtr(PLANAR_V),
			src->GetPitch(PLANAR_V), dst->GetPitch(PLANAR_V),
			row_size>>1, height>>1, Cth, divin, strength);

	return dst;
}


AVSValue __cdecl Create_SmoothIQ(AVSValue args, void* user_data, IScriptEnvironment* env) 
{
	int 	radiusC = 3, thresC = 270, strength = 255;
	bool	field = false, hqy = false, hqc = false;
		  
	return new SSmooth(
		args[0].AsClip(),
		1,							//Doesn't apply anyway
		args[1].AsInt(radiusC),
		-1337,						//Doesn't apply anyway
		args[2].AsInt(thresC),
		255,						//Doesn't apply anyway
		false,
		false,
		args[3].AsBool(field),
		env);
}

AVSValue __cdecl Create_SSHiQ(AVSValue args, void* user_data, IScriptEnvironment* env) 
{
	int 	radiusY = 5, radiusC = 3;
	int		thresY = 20, thresC = 30;
	int		strength = 240;
	bool	hqy = true, hqc = true, interlaced = false;
		  
	return new SSmooth(
		args[0].AsClip(),
		args[1].AsInt(radiusY),
		args[2].AsInt(radiusC),
		args[3].AsInt(thresY),
		args[4].AsInt(thresC),
		args[5].AsInt(strength),
		args[6].AsBool(hqy),
		args[7].AsBool(hqc),
		args[8].AsBool(interlaced),
		env);
}

extern "C" __declspec(dllexport) const char* __stdcall
AvisynthPluginInit2(IScriptEnvironment* env) {
	env->AddFunction("SmoothUV", "c[radius]i[T]i[field]b", Create_SmoothIQ, 0);
	env->AddFunction("SSHiQ", "c[rY]i[rC]i[tY]i[tC]i[str]i[HQY]b[HQC]b[field]b",
					Create_SSHiQ, 0);
	return "`SmartSmoother' spatial denoiser and derainbower";
}
